File: Difficulty: Easy/Inorder Traversal/inorder-traversal.cpp
/*
// Tree Node
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    // Constructor to initialize a new node
    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};
*/

class Solution {
  public:
    // Function to return a list containing the inorder traversal of the tree.
    void dfs(Node* root, vector<int>& ans) {
        if(root==nullptr) return;
        dfs(root->left,ans);
        ans.push_back(root->data);
        dfs(root->right,ans);
    }
    
    vector<int> inOrder(Node* root) {
        // Your code here
        vector<int> ans;
        dfs(root,ans);
        return ans;
    }
};

File: Difficulty: Easy/Inorder Traversal/inorder-traversal.java
/*
class Node {
    int data;
    Node left, right;
    Node(int item){
        data = item;
        left = right = null;
    }
}
*/
class Solution {
    public ArrayList<Integer> inOrder(Node root) {
        // code here
        Deque<Node> st = new ArrayDeque<>();
    
        List<Integer> arr = new ArrayList<>();
        Node cur = root;
        
        while(true) {
            
            if(cur!=null) {
                st.push(cur);
                cur = cur.left;
            }
            else {
                Node temp = st.poll();
                arr.add(temp.data);
                
                cur = temp.right;
                if(cur!=null) {
                    st.push(cur);
                    cur = cur.left;
                }
               
                
            }
            if(st.isEmpty()==true) break;
        }
        return (ArrayList<Integer>)arr;
    }
}